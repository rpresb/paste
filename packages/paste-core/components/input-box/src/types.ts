import * as PropTypes from 'prop-types';
import type {BoxProps} from '@twilio-paste/box';

export type InputBoxTypes = 'text' | 'email' | 'hidden' | 'number' | 'password' | 'search' | 'tel' | 'date' | 'time';
export type Variants = 'default' | 'inverse';

export interface FauxInputProps extends Pick<BoxProps, 'element'> {
  children: NonNullable<React.ReactNode>;
  disabled?: boolean;
  hasError?: boolean;
  readOnly?: boolean;
  type?: InputBoxTypes;
  variant?: Variants;
}

export const FauxInputPropTypes = {
  children: PropTypes.node.isRequired,
  disabled: PropTypes.bool,
  element: PropTypes.string,
  hasError: PropTypes.bool,
  readOnly: PropTypes.bool,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  type: PropTypes.oneOf(['text', 'email', 'hidden', 'number', 'password', 'search', 'tel', 'date', 'time']) as any,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  variant: PropTypes.oneOf(['default', 'inverse']) as any,
};
