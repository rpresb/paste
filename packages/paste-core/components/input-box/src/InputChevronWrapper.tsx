import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box} from '@twilio-paste/box';
import type {BoxProps} from '@twilio-paste/box';

export interface InputChevronWrapperProps extends Pick<BoxProps, 'element'> {
  children: NonNullable<React.ReactNode>;
}

export const InputChevronWrapper = React.forwardRef<HTMLDivElement, InputChevronWrapperProps>(
  ({children, element}, ref) => (
    <Box
      alignItems="center"
      display="inline-flex"
      position="absolute"
      pointerEvents="none"
      right="space30"
      top="50%"
      transform="translateY(-50%)"
      zIndex="zIndex10"
      element={`${element}_CHEVRON_WRAPPER`}
      ref={ref}
    >
      {children}
    </Box>
  )
);

InputChevronWrapper.displayName = 'InputChevronWrapper';

InputChevronWrapper.propTypes = {
  children: PropTypes.node.isRequired,
  element: PropTypes.string,
};
