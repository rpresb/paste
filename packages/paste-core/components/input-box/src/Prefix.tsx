import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box} from '@twilio-paste/box';
import type {BoxProps} from '@twilio-paste/box';
import type {BackgroundColor, BorderColor} from '@twilio-paste/style-props';
import type {Variants} from './types';

export interface PrefixProps extends Pick<BoxProps, 'element'> {
  children: NonNullable<React.ReactNode>;
  disabled?: boolean;
  variant?: Variants;
}

const Prefix = React.forwardRef<HTMLDivElement, PrefixProps>(({children, disabled, element, variant}, ref) => {
  let backgroundColor = 'colorBackground' as BackgroundColor;
  let borderColor = 'colorBorderWeaker' as BorderColor;
  if (disabled && variant === 'inverse') {
    backgroundColor = 'none';
    borderColor = 'colorBorderInverseWeaker';
  } else if (variant === 'inverse') {
    backgroundColor = 'colorBackgroundInverseStrong';
    borderColor = 'colorBorderInverse';
  }

  if (children == null) return null;
  return (
    <Box
      alignItems="flex-start"
      backgroundColor={backgroundColor}
      borderBottomLeftRadius="borderRadius20"
      borderRightColor={borderColor}
      borderRightStyle="solid"
      borderRightWidth="borderWidth10"
      borderTopLeftRadius="borderRadius20"
      display="flex"
      element={`${element}_PREFIX`}
      lineHeight="lineHeight20"
      padding="space30"
      ref={ref}
    >
      {children}
    </Box>
  );
});

Prefix.displayName = 'Prefix';

Prefix.propTypes = {
  children: PropTypes.node.isRequired,
  element: PropTypes.string,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  variant: PropTypes.oneOf(['default', 'inverse']) as any,
};

export {Prefix};
