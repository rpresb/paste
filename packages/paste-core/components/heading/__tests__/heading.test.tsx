import * as React from 'react';
import {render, screen} from '@testing-library/react';
import {Theme} from '@twilio-paste/theme';
import {CustomizationProvider} from '@twilio-paste/customization';
import {matchers} from 'jest-emotion';
// @ts-ignore typescript doesn't like js imports
import axe from '../../../../../.jest/axe-helper';
import {Heading} from '../src';

expect.extend(matchers);

describe('Heading', () => {
  describe('Render', () => {
    it('should render an H1 at fontSize90', (): void => {
      const {asFragment} = render(
        <Theme.Provider theme="default">
          <Heading as="h1" variant="heading10">
            This is an H1
          </Heading>
        </Theme.Provider>
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render an H2 at fontSize70', (): void => {
      const {asFragment} = render(
        <Theme.Provider theme="default">
          <Heading as="h2" variant="heading20">
            This is an H2
          </Heading>
        </Theme.Provider>
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render an H3 at fontSize60', (): void => {
      const {asFragment} = render(
        <Theme.Provider theme="default">
          <Heading as="h3" variant="heading30">
            This is an H3
          </Heading>
        </Theme.Provider>
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render an H4 at fontSize40', (): void => {
      const {asFragment} = render(
        <Theme.Provider theme="default">
          <Heading as="h4" variant="heading40">
            This is an H4
          </Heading>
        </Theme.Provider>
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render an H5 at fontSize30', (): void => {
      const {asFragment} = render(
        <Theme.Provider theme="default">
          <Heading as="h5" variant="heading50">
            This is an H5
          </Heading>
        </Theme.Provider>
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render an H6 at fontSize20', (): void => {
      const {asFragment} = render(
        <Theme.Provider theme="default">
          <Heading as="h6" variant="heading60">
            This is an H6
          </Heading>
        </Theme.Provider>
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render an italic H2 at fontSize50', (): void => {
      const {asFragment} = render(
        <Theme.Provider theme="default">
          <Heading as="h2" variant="heading20">
            <i>This is an italic H2</i>
          </Heading>
        </Theme.Provider>
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render with no margin', (): void => {
      const {asFragment} = render(
        <Theme.Provider theme="default">
          <Heading as="h2" marginBottom="space0" variant="heading10">
            no margin heading
          </Heading>
          <Heading as="h2" marginBottom="space0" variant="heading20">
            no margin heading
          </Heading>
          <Heading as="h2" marginBottom="space0" variant="heading30">
            no margin heading
          </Heading>
          <Heading as="h2" marginBottom="space0" variant="heading40">
            no margin heading
          </Heading>
          <Heading as="h2" marginBottom="space0" variant="heading50">
            no margin heading
          </Heading>
          <Heading as="h2" marginBottom="space0" variant="heading60">
            no margin heading
          </Heading>
        </Theme.Provider>
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('HTML attributes', () => {
    it('should set a element data attribute for Heading', () => {
      render(
        <Heading as="h1" variant="heading10">
          heading
        </Heading>
      );
      expect(screen.getByRole('heading').getAttribute('data-paste-element')).toEqual('HEADING');
    });
    it('should set a custom element data attribute for Heading', () => {
      render(
        <Heading as="h1" variant="heading10" element="foo">
          heading
        </Heading>
      );
      expect(screen.getByRole('heading').getAttribute('data-paste-element')).toEqual('foo');
    });
  });

  describe('Customization', () => {
    it('should add custom styles to Heading', (): void => {
      render(
        <CustomizationProvider
          baseTheme="default"
          elements={{HEADING: {color: 'colorTextWeak', backgroundColor: 'colorBackground'}}}
        >
          <Heading as="h1" variant="heading10">
            Custom heading
          </Heading>
        </CustomizationProvider>
      );
      const renderedHeading = screen.getByRole('heading');
      expect(renderedHeading).toHaveStyleRule('background-color', 'rgb(244,244,246)');
      expect(renderedHeading).toHaveStyleRule('color', 'rgb(96,107,133)');
    });

    it('should add custom styles to Heading with a custom element data attribute', (): void => {
      render(
        <CustomizationProvider
          baseTheme="default"
          elements={{foo: {color: 'colorTextWeak', backgroundColor: 'colorBackground'}}}
        >
          <Heading as="h1" variant="heading10" element="foo">
            Custom heading
          </Heading>
        </CustomizationProvider>
      );
      const renderedHeading = screen.getByRole('heading');
      expect(renderedHeading).toHaveStyleRule('background-color', 'rgb(244,244,246)');
      expect(renderedHeading).toHaveStyleRule('color', 'rgb(96,107,133)');
    });
  });

  describe('Accessibility', () => {
    it('should have no accessibility violations', async () => {
      const {container} = render(
        <Theme.Provider theme="default">
          <Heading as="h1" variant="heading10">
            This is an H1
          </Heading>
          <Heading as="h2" variant="heading20">
            This is an H2
          </Heading>
          <Heading as="h3" variant="heading30">
            This is an H3
          </Heading>
          <Heading as="h4" variant="heading40">
            This is an H4
          </Heading>
          <Heading as="h5" variant="heading50">
            This is an H5
          </Heading>
          <Heading as="h6" variant="heading60">
            This is an H6
          </Heading>
        </Theme.Provider>
      );
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });
});
