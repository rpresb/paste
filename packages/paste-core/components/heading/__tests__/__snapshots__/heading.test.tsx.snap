// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading Render should render an H1 at fontSize90 1`] = `
<DocumentFragment>
  .emotion-1 {
  color: rgb(18,28,45);
  font-size: 0.875rem;
  font-family: 'Inter var experimental','Inter var',-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
  font-variant-numeric: tabular-nums;
}

.emotion-0 {
  margin: 0;
  padding: 0;
  color: rgb(18,28,45);
  font-size: 2rem;
  line-height: 2.75rem;
  margin-bottom: 1.5rem;
  font-weight: 700;
  display: block;
}

<div
    class="emotion-1"
  >
    <h1
      class="emotion-0"
      color="colorText"
      data-paste-element="HEADING"
      display="block"
      font-size="fontSize90"
      font-weight="fontWeightBold"
    >
      This is an H1
    </h1>
  </div>
</DocumentFragment>
`;

exports[`Heading Render should render an H2 at fontSize70 1`] = `
<DocumentFragment>
  .emotion-1 {
  color: rgb(18,28,45);
  font-size: 0.875rem;
  font-family: 'Inter var experimental','Inter var',-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
  font-variant-numeric: tabular-nums;
}

.emotion-0 {
  margin: 0;
  padding: 0;
  color: rgb(18,28,45);
  font-size: 1.5rem;
  line-height: 2rem;
  margin-bottom: 1.25rem;
  font-weight: 700;
  display: block;
}

<div
    class="emotion-1"
  >
    <h2
      class="emotion-0"
      color="colorText"
      data-paste-element="HEADING"
      display="block"
      font-size="fontSize70"
      font-weight="fontWeightBold"
    >
      This is an H2
    </h2>
  </div>
</DocumentFragment>
`;

exports[`Heading Render should render an H3 at fontSize60 1`] = `
<DocumentFragment>
  .emotion-1 {
  color: rgb(18,28,45);
  font-size: 0.875rem;
  font-family: 'Inter var experimental','Inter var',-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
  font-variant-numeric: tabular-nums;
}

.emotion-0 {
  margin: 0;
  padding: 0;
  color: rgb(18,28,45);
  font-size: 1.25rem;
  line-height: 1.75rem;
  margin-bottom: 1rem;
  font-weight: 700;
  display: block;
}

<div
    class="emotion-1"
  >
    <h3
      class="emotion-0"
      color="colorText"
      data-paste-element="HEADING"
      display="block"
      font-size="fontSize60"
      font-weight="fontWeightBold"
    >
      This is an H3
    </h3>
  </div>
</DocumentFragment>
`;

exports[`Heading Render should render an H4 at fontSize40 1`] = `
<DocumentFragment>
  .emotion-1 {
  color: rgb(18,28,45);
  font-size: 0.875rem;
  font-family: 'Inter var experimental','Inter var',-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
  font-variant-numeric: tabular-nums;
}

.emotion-0 {
  margin: 0;
  padding: 0;
  color: rgb(18,28,45);
  font-size: 1rem;
  line-height: 1.5rem;
  margin-bottom: 0.75rem;
  font-weight: 700;
  display: block;
}

<div
    class="emotion-1"
  >
    <h4
      class="emotion-0"
      color="colorText"
      data-paste-element="HEADING"
      display="block"
      font-size="fontSize40"
      font-weight="fontWeightBold"
    >
      This is an H4
    </h4>
  </div>
</DocumentFragment>
`;

exports[`Heading Render should render an H5 at fontSize30 1`] = `
<DocumentFragment>
  .emotion-1 {
  color: rgb(18,28,45);
  font-size: 0.875rem;
  font-family: 'Inter var experimental','Inter var',-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
  font-variant-numeric: tabular-nums;
}

.emotion-0 {
  margin: 0;
  padding: 0;
  color: rgb(18,28,45);
  font-size: 0.875rem;
  line-height: 1.25rem;
  margin-bottom: 0.5rem;
  font-weight: 700;
  display: block;
}

<div
    class="emotion-1"
  >
    <h5
      class="emotion-0"
      color="colorText"
      data-paste-element="HEADING"
      display="block"
      font-size="fontSize30"
      font-weight="fontWeightBold"
    >
      This is an H5
    </h5>
  </div>
</DocumentFragment>
`;

exports[`Heading Render should render an H6 at fontSize20 1`] = `
<DocumentFragment>
  .emotion-1 {
  color: rgb(18,28,45);
  font-size: 0.875rem;
  font-family: 'Inter var experimental','Inter var',-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
  font-variant-numeric: tabular-nums;
}

.emotion-0 {
  margin: 0;
  padding: 0;
  color: rgb(18,28,45);
  font-size: 0.75rem;
  line-height: 1.25rem;
  margin-bottom: 0.5rem;
  font-weight: 700;
  display: block;
}

<div
    class="emotion-1"
  >
    <h6
      class="emotion-0"
      color="colorText"
      data-paste-element="HEADING"
      display="block"
      font-size="fontSize20"
      font-weight="fontWeightBold"
    >
      This is an H6
    </h6>
  </div>
</DocumentFragment>
`;

exports[`Heading Render should render an italic H2 at fontSize50 1`] = `
<DocumentFragment>
  .emotion-1 {
  color: rgb(18,28,45);
  font-size: 0.875rem;
  font-family: 'Inter var experimental','Inter var',-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
  font-variant-numeric: tabular-nums;
}

.emotion-0 {
  margin: 0;
  padding: 0;
  color: rgb(18,28,45);
  font-size: 1.5rem;
  line-height: 2rem;
  margin-bottom: 1.25rem;
  font-weight: 700;
  display: block;
}

<div
    class="emotion-1"
  >
    <h2
      class="emotion-0"
      color="colorText"
      data-paste-element="HEADING"
      display="block"
      font-size="fontSize70"
      font-weight="fontWeightBold"
    >
      <i>
        This is an italic H2
      </i>
    </h2>
  </div>
</DocumentFragment>
`;

exports[`Heading Render should render with no margin 1`] = `
<DocumentFragment>
  .emotion-6 {
  color: rgb(18,28,45);
  font-size: 0.875rem;
  font-family: 'Inter var experimental','Inter var',-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
  font-variant-numeric: tabular-nums;
}

.emotion-0 {
  margin: 0;
  padding: 0;
  color: rgb(18,28,45);
  font-size: 2rem;
  line-height: 2.75rem;
  margin-bottom: 0;
  font-weight: 700;
  display: block;
}

.emotion-1 {
  margin: 0;
  padding: 0;
  color: rgb(18,28,45);
  font-size: 1.5rem;
  line-height: 2rem;
  margin-bottom: 0;
  font-weight: 700;
  display: block;
}

.emotion-2 {
  margin: 0;
  padding: 0;
  color: rgb(18,28,45);
  font-size: 1.25rem;
  line-height: 1.75rem;
  margin-bottom: 0;
  font-weight: 700;
  display: block;
}

.emotion-3 {
  margin: 0;
  padding: 0;
  color: rgb(18,28,45);
  font-size: 1rem;
  line-height: 1.5rem;
  margin-bottom: 0;
  font-weight: 700;
  display: block;
}

.emotion-4 {
  margin: 0;
  padding: 0;
  color: rgb(18,28,45);
  font-size: 0.875rem;
  line-height: 1.25rem;
  margin-bottom: 0;
  font-weight: 700;
  display: block;
}

.emotion-5 {
  margin: 0;
  padding: 0;
  color: rgb(18,28,45);
  font-size: 0.75rem;
  line-height: 1.25rem;
  margin-bottom: 0;
  font-weight: 700;
  display: block;
}

<div
    class="emotion-6"
  >
    <h2
      class="emotion-0"
      color="colorText"
      data-paste-element="HEADING"
      display="block"
      font-size="fontSize90"
      font-weight="fontWeightBold"
    >
      no margin heading
    </h2>
    <h2
      class="emotion-1"
      color="colorText"
      data-paste-element="HEADING"
      display="block"
      font-size="fontSize70"
      font-weight="fontWeightBold"
    >
      no margin heading
    </h2>
    <h2
      class="emotion-2"
      color="colorText"
      data-paste-element="HEADING"
      display="block"
      font-size="fontSize60"
      font-weight="fontWeightBold"
    >
      no margin heading
    </h2>
    <h2
      class="emotion-3"
      color="colorText"
      data-paste-element="HEADING"
      display="block"
      font-size="fontSize40"
      font-weight="fontWeightBold"
    >
      no margin heading
    </h2>
    <h2
      class="emotion-4"
      color="colorText"
      data-paste-element="HEADING"
      display="block"
      font-size="fontSize30"
      font-weight="fontWeightBold"
    >
      no margin heading
    </h2>
    <h2
      class="emotion-5"
      color="colorText"
      data-paste-element="HEADING"
      display="block"
      font-size="fontSize20"
      font-weight="fontWeightBold"
    >
      no margin heading
    </h2>
  </div>
</DocumentFragment>
`;
